### ISSUE TITLE
compat library installed but not loaded

### ISSUE THREAD

**Issue by gan-ainm** [SystemCrafters/crafted-emacs#282]
When I do the initial invocation of Crafted Emacs (via Chemacs2 `/usr/bin/emacs --with-profile crafted`), it automatically installs the **compat** library:

```
Loading /home/gan/.local/share/crafted-emacs/modules/crafted-early-init-config.el (source)...done
crafted-package-config: checking package archives
obsolete timestamp with cdr 1 [10 times]
crafted-package-config: package archives stale, refreshing
Importing package-keyring.gpg...done
crafted-package-config: package system initialized!
Loading /home/gan/.local/share/crafted-emacs/modules/crafted-init-config.el (source)...
Contacting host: elpa.gnu.org:443
Parsing tar file...done
Extracting...done
  INFO     Scraping files for compat-autoloads.el... 
Generating autoloads for compat-25.el...done
Generating autoloads for compat-26.el...done
Generating autoloads for compat-27.el...done
Generating autoloads for compat-28.el...done
Generating autoloads for compat-29.el...done
Generating autoloads for compat-macs.el...done
Generating autoloads for compat-pkg.el...done
Generating autoloads for compat.el...done
  INFO     Scraping files for compat-autoloads.el...done
Wrote /home/gan/.local/share/crafted-emacs/elpa/compat-29.1.4.1/compat-autoloads.el [2 times]
Checking /home/gan/.local/share/crafted-emacs/elpa/compat-29.1.4.1...
Wrote /home/gan/.local/share/crafted-emacs/elpa/compat-29.1.4.1/compat-25.elc
Checking /home/gan/.local/share/crafted-emacs/elpa/compat-29.1.4.1...
Wrote /home/gan/.local/share/crafted-emacs/elpa/compat-29.1.4.1/compat-26.elc
Checking /home/gan/.local/share/crafted-emacs/elpa/compat-29.1.4.1...
Wrote /home/gan/.local/share/crafted-emacs/elpa/compat-29.1.4.1/compat-27.elc
Checking /home/gan/.local/share/crafted-emacs/elpa/compat-29.1.4.1...
Wrote /home/gan/.local/share/crafted-emacs/elpa/compat-29.1.4.1/compat-28.elc
Checking /home/gan/.local/share/crafted-emacs/elpa/compat-29.1.4.1...
Wrote /home/gan/.local/share/crafted-emacs/elpa/compat-29.1.4.1/compat-29.elc
Checking /home/gan/.local/share/crafted-emacs/elpa/compat-29.1.4.1... [4 times]
Wrote /home/gan/.local/share/crafted-emacs/elpa/compat-29.1.4.1/compat.elc
Checking /home/gan/.local/share/crafted-emacs/elpa/compat-29.1.4.1...
Done (Total of 6 files compiled, 3 skipped)
Package ‘compat’ installed.
adding modules to load-path: /home/gan/.local/share/crafted-emacs/modules/
Loading /home/gan/.local/share/crafted-emacs/modules/crafted-init-config.el (source)...done
Repeat mode is enabled for 14 commands and 7 keymaps; see ‘describe-repeat-maps’
Cleaning up the recentf list...done (0 removed)
(New file)
Package refresh done
Perform Emacs Lisp header auto-insertion? (y or n) y
Saving file /home/gan/.local/share/crafted-emacs/custom.el...
Wrote /home/gan/.local/share/crafted-emacs/custom.el
Updating buffer list...
Formats have changed, recompiling...done
Updating buffer list...done
```
When launching Crafted Emacs the second time, the messages are:
```
Loading /home/gan/.local/share/crafted-emacs/modules/crafted-early-init-config.el (source)...done
crafted-package-config: checking package archives
obsolete timestamp with cdr 1 [10 times]
crafted-package-config: package archives stale, refreshing
Importing package-keyring.gpg...done
crafted-package-config: package system initialized!
Loading /home/gan/.local/share/crafted-emacs/modules/crafted-init-config.el (source)...
‘compat’ is already installed
adding modules to load-path: /home/gan/.local/share/crafted-emacs/modules/
Loading /home/gan/.local/share/crafted-emacs/modules/crafted-init-config.el (source)...done
Loading /home/gan/.local/share/crafted-emacs/recentf...done
Cleaning up the recentf list...done (0 removed)
Saving file /home/gan/.local/share/crafted-emacs/custom.el...
Wrote /home/gan/.local/share/crafted-emacs/custom.el
Saving file /home/gan/.local/share/crafted-emacs/custom.el...
Wrote /home/gan/.local/share/crafted-emacs/custom.el
For information about GNU Emacs and the GNU system, type C-h C-a.
Package refresh done
```
The warning that pops up:
```
Warning (initialization): An error occurred while loading ‘/home/gan/.config/emacs/init.el’:

Symbol's function definition is void: keymap-set

To ensure normal operation, you should investigate and remove the
cause of the error in your initialization file.  Start Emacs with
the ‘--debug-init’ option to view a complete error backtrace. Disable showing Disable logging
```
And here is the start debugger output:
```
Debugger entered--Lisp error: (void-function keymap-set)
  (keymap-set 'crafted-windows-key-map "u" 'winner-undo)
  load-with-code-conversion("/home/gan/.local/share/crafted-emacs/modules/craft..." "/home/gan/.local/share/crafted-emacs/modules/craft..." nil t)
  require(crafted-defaults-config)
  load-with-code-conversion("/home/gan/.local/share/crafted-emacs/init.el" "/home/gan/.local/share/crafted-emacs/init.el" t t)
  load("/home/gan/.local/share/crafted-emacs/init.el" t t)
  (let ((init-file (expand-file-name "init.el" user-emacs-directory))) (setq package-user-dir (expand-file-name "elpa" user-emacs-directory)) (load init-file t t) (let ((chemacs-custom-file (chemacs-profile-get 'custom-file init-file))) (if (not custom-file) (progn (setq custom-file chemacs-custom-file) (if (equal custom-file init-file) nil (if (file-exists-p custom-file) nil (let (...) (save-current-buffer ... ...))) (load custom-file))))))
  chemacs-load-user-init()
  load-with-code-conversion("/home/gan/.config/emacs/init.el" "/home/gan/.config/emacs/init.el" t t)
  load("/home/gan/.config/emacs/init" noerror nomessage)
  startup--load-user-init-file(#f(compiled-function () #<bytecode 0x1a0964c7f16d38b4>) #f(compiled-function () #<bytecode -0x6201fff00760d94>) t)
  command-line()
  normal-top-level()
```
And here is the loadpath:
```
load-path is a variable defined in ‘C source code’.

Its value is
("/home/gan/.local/share/crafted-emacs/modules/" "/usr/share/emacs/28.2/site-lisp" "/usr/share/emacs/site-lisp" "/usr/share/emacs/site-lisp/auctex" "/usr/share/emacs/site-lisp/mu4e" "/usr/share/emacs/site-lisp/slime" "/usr/share/emacs/site-lisp/themes" "/usr/share/emacs/site-lisp/auctex/images" "/usr/share/emacs/site-lisp/slime/contrib" "/usr/share/emacs/site-lisp/slime/doc" "/usr/share/emacs/site-lisp/slime/swank" "/usr/share/emacs/site-lisp/slime/contrib/test" "/usr/share/emacs/28.2/lisp" "/usr/share/emacs/28.2/lisp/vc" "/usr/share/emacs/28.2/lisp/url" "/usr/share/emacs/28.2/lisp/textmodes" "/usr/share/emacs/28.2/lisp/progmodes" "/usr/share/emacs/28.2/lisp/play" "/usr/share/emacs/28.2/lisp/org" "/usr/share/emacs/28.2/lisp/nxml" "/usr/share/emacs/28.2/lisp/net" "/usr/share/emacs/28.2/lisp/mh-e" "/usr/share/emacs/28.2/lisp/mail" "/usr/share/emacs/28.2/lisp/leim" "/usr/share/emacs/28.2/lisp/language" "/usr/share/emacs/28.2/lisp/international" "/usr/share/emacs/28.2/lisp/image" "/usr/share/emacs/28.2/lisp/gnus" "/usr/share/emacs/28.2/lisp/eshell" "/usr/share/emacs/28.2/lisp/erc" "/usr/share/emacs/28.2/lisp/emulation" "/usr/share/emacs/28.2/lisp/emacs-lisp" "/usr/share/emacs/28.2/lisp/cedet" "/usr/share/emacs/28.2/lisp/calendar" "/usr/share/emacs/28.2/lisp/calc" "/usr/share/emacs/28.2/lisp/obsolete")

```
Starting Emacs with the compat directory explicitly added works: `EMACSLOADPATH=/home/gan/.local/share/crafted-emacs/elpa/compat-29.1.4.1: /usr/bin/emacs --with-profile crafted`

**Comment by jeffbowman** [comment]
Thanks for this detailed report!!   This is very helpful!

**Comment by jeffbowman** [comment]
Still looking into this.  I'm not seeing the problem in my clean room VM, so this may be an issue when using `chemacs2`.

**Comment by jeffbowman** [comment]
Even in chemacs2, I can't duplicate this error problem.  Can you please provide the init.el and early-init.el you are using for this profile?  Then I will check with those files instead of the example files.

**Comment by gan-ainm** [comment]
I must admit, that my _init.el_ and _early-init.el_ reside within `crafted-emacs-home`, i.e., it is colocated with `user-emacs-home` as defined in the Chemacs2 profile. Can that be an issue?

Here's _init.el_:
```
(setq custom-file (expand-file-name "custom.el" user-emacs-directory))
(when (and custom-file
           (file-exists-p custom-file))
  (load custom-file nil :nomessage))

(load "~/.local/share/crafted-emacs/modules/crafted-init-config")

(require 'crafted-defaults-config)
```
and _early-init.el_:
```
(load "~/.local/share/crafted-emacs/modules/crafted-early-init-config")
```
and the Chemacs2 _profiles.el_:
```
(("spacemacs" . ((user-emacs-directory . "~/.config/spacemacs")
                 (server-name . "space")))
 ("doom" . ((user-emacs-directory . "~/.local/share/doom-emacs")
            (env . (("DOOMDIR" . "~/.config/doom")))
            (server-name . "doom")))
 ("scratch" . ((user-emacs-directory . "~/.config/Emacs_from_Scratch")
               (server-name . "scratch")))
 ("efs" . ((user-emacs-directory . "~/Projects/efs")
           (server-name . "efs")))
 ("rational" . ((user-emacs-directory . "~/.config/rational-emacs")
                (env . (("RATIONAL_EMACS_HOME" . "~/.config/rational-emacs/config")))
                (server-name . "rational")))
 ("mastering" . ((user-emacs-directory . "~/.config/mastering-emacs")
                (server-name . "mastering")))
 ("glickstein" . ((user-emacs-directory . "~/.config/glickstein")
                (server-name . "glickstein")))
 ("raw" . ((user-emacs-directory . "~/.config/raw")
           (server-name . "raw")))
 ("crafted" . ((user-emacs-directory . "~/.local/share/crafted-emacs")
               (server-name . "crafted"))))
```

**Comment by jeffbowman** [comment]
> I must admit, that my init.el and early-init.el reside within crafted-emacs-home, i.e., it is colocated with user-emacs-home as defined in the Chemacs2 profile. Can that be an issue?

That shouldn't matter.  I'll try setting mine up that way and see if it does.  Also, are you running multiple Emacs daemons?

**Comment by jeffbowman** [comment]
So, I setup a new VM just to have a completely new clean room.  Then I did the following:

* mkdir -p .local/share
* cd .local/share
* git clone https://github.com/systemcrafters/crafted-emacs
* git checkout craftedv2beta
* edit init.el and early-init.el as you have them above
* cd ~
* clone chemacs2
* edit .emacs-profiles.el and use your "crafted" profile as my "default" profile
* start Emacs - my `*Messages*` buffer looks like your example
* restart emacs (ie C-x C-c, then run `emacs` again) however, I'm not seeing the same kind of `*Messages*`. as you have posted.
* switch to the `*scratch*` buffer and run `(fboundp 'keymap-set)` which returns `t`, also `(featurep 'compat)` returns `t` and `(featurep 'compat-29)` returns `t`. 

So.... not sure I'm starting Emacs like this: `emacs --with-profile default` or like this: `emacs --with-profile default --fg-daemon=crafted` and I'm not seeing any issues.

If you think I've missed something, please point it out and I'll try again.

**Comment by gan-ainm** [comment]
Oh, I just realized that they must have switched Chemacs2's default branch from **master** to **main** some time ago and all my pull efforts were fruitless since May 2021. I checked out **main** and everythings works as described in your most recent post.

Thanks a lot for your efforts an sorry for stealing your precious time!

BTW: I already started my first _custom-org-config.el_ module based on the _crafted-org-config.el_.
